### PR TITLE
Adapt to LLVM 16

### PR DESCRIPTION
**Pull request by vchuravy**
I love C++ so much...

starts the slog towards upgrading us to LLVM 16 and C++17.

### CODE REVIEW

File: Make.inc [hunk 1]
```diff
@@ -497,7 +497,7 @@ JCPPFLAGS_COMMON  := -fasynchronous-unwind-tables
 JCPPFLAGS_CLANG   := $(JCPPFLAGS_COMMON) -mllvm -enable-tail-merge=0
 JCPPFLAGS_GCC     := $(JCPPFLAGS_COMMON) -fno-tree-tail-merge
 
-JCXXFLAGS_COMMON  := -pipe $(fPIC) -fno-rtti -std=c++14
+JCXXFLAGS_COMMON  := -pipe $(fPIC) -fno-rtti -std=c++17
```
Comment (giordano) [inline Make.inc:500]: Does this change the minimum version of compilers required to build julia?

Comment (imciner2) [reply Make.inc:500]: If the only new thing is the usage of `std::optional` here, then I think it only raises the baseline for Apple Clang to be Apple Clang 10.0 instead of 9.3. `std::optional` should already be present in GCC 7.1 and Clang 5.0, which are currently the listed minimums.

That said, LLVM says they want GCC 7.4 minimum (and Apple Clang 10), so for a local build of LLVM the GCC minimum will be higher than what we currently say.

**Comment by Robert-j7** [comment]
One general question, why not just skip LLVM 16 and go straight to LLVM 17?

**Comment by giordano** [comment]
> One general question, why not just skip LLVM 16 and go straight to LLVM 17? 

"just" and "straight" are easier said than done: [we started the effort to build llvm 16](https://github.com/JuliaPackaging/Yggdrasil/pull/6777) 5 months ago, and that's not even fully done yet.

**Comment by vchuravy** [comment]
> One general question, why not just skip LLVM 16 and go straight to LLVM 17?

Are you volunteering to do that work? LLVM updates are tedious, fraught with risk and while it might be nice to skip an entire version you end up with more work and less ability to find out when something broke, because you can't bisect the middle.

If anyone wants to get involve getting the LLVM 16 build into Yggdrasil and looking at LLVM 17 in Yggdrasil would be a major contribution.

**Comment by vchuravy** [comment]
Okay two remaining source files...

```
make[1]: *** [/home/vchuravy/src/julia/src/Makefile:240: pipeline.dbg.obj] Error 1
make[1]: *** [/home/vchuravy/src/julia/src/Makefile:240: jitlayers.dbg.obj] Error 1
```

The pipeline failures are:

```
/home/vchuravy/src/julia/src/pipeline.cpp: In function 'void buildEarlySimplificationPipeline(llvm::ModulePassManager&, llvm::PassBuilder*, llvm::OptimizationLevel, const OptimizationOptions&)':
/home/vchuravy/src/julia/src/pipeline.cpp:345:33: error: use of deleted function 'llvm::PassManager<llvm::Module>::PassManager(const llvm::PassManager<llvm::Module>&)'
  345 |     invokePipelineStartCallbacks(MPM, PB, O);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
```

Which makes me think that we are missing one of @pchintalapudi passes from LLVM 15.

**Comment by vchuravy** [comment]
Okay @pchintalapudi this is probably ready for a first review.

definitly didn't do the memory attribute transition correctly since:

```
Attribute 'readnone' does not apply to functions!
  %10 = call nonnull ptr @julia.pointer_from_objref(ptr addrspace(11) %9) #3
Attribute 'readnone' does not apply to functions!
  %33 = call nonnull ptr @julia.pointer_from_objref(ptr addrspace(11) %32) #3
Attribute 'readnone' does not apply to functions!
```

**Comment by vchuravy** [comment]
```
cat deps/llvm.version
# -*- makefile -*-

## jll artifact
LLVM_JLL_NAME := libLLVM
LLVM_ASSERT_JLL_VER := 15.0.7+8
## source build
# Version number of LLVM
LLVM_VER := 16.0.6
# Git branch name in `LLVM_GIT_URL` repository
LLVM_BRANCH=julia-16.0.6-0
# Git ref in `LLVM_GIT_URL` repository
LLVM_SHA1=julia-16.0.6-0

## Following options are used to automatically fetch patchset from Julia's fork.  This is
## useful if you want to build an external LLVM while still applying Julia's patches.
# Set to 1 if you want to automatically apply Julia's patches to a different fork of LLVM.
LLVM_APPLY_JULIA_PATCHES := 0
# GitHub repository to use for fetching the Julia patches to apply to LLVM source code.
LLVM_JULIA_DIFF_GITHUB_REPO := https://github.com/llvm/llvm-project
# Base GitHub ref for generating the diff.
LLVM_BASE_REF := llvm:llvmorg-16.0.6
# Julia fork's GitHub ref for generating the diff.
LLVM_JULIA_REF := JuliaLang:julia-16.0.6-0
```

with

```
cat Make.user
USE_BINARYBUILDER_LLVM:=0
DEPS_GIT=1

FORCE_ASSERTIONS=1
LLVM_ASSERTIONS=1

LLVM_DEBUG=2
override JULIA_BUILD_MODE=debug
```

File: src/codegen.cpp [hunk 2]
```diff
@@ -7559,10 +7672,16 @@ static jl_llvm_functions_t
     // Step 4b. determine debug info signature and other type info for locals
     DICompileUnit::DebugEmissionKind emissionKind = (DICompileUnit::DebugEmissionKind) ctx.params->debug_info_kind;
     DICompileUnit::DebugNameTableKind tableKind;
+#if JL_LLVM_VERSION >= 160000
+#undef None
```
Comment (pchintalapudi) [inline src/codegen.cpp:7676]: IMO we should just define `template<typename T> using Optional = std::optional<T>` and replace the explicit function calls with the ones supported by both implementations (e.g. operator bool, operator*, etc). That'll reduce the diff size significantly.

File: src/ccall.cpp [hunk 3]
```diff
@@ -1686,7 +1695,11 @@ static jl_cgval_t emit_ccall(jl_codectx_t &ctx, jl_value_t **args, size_t nargs)
                 true);
         setName(ctx.emission_context, signal_page_load, "signal_page_load");
         ctx.builder.CreateBr(contBB);
+#if JL_LLVM_VERSION >= 160000
+        ctx.f->insert(ctx.f->end(), contBB);
```
Comment (pchintalapudi) [inline src/ccall.cpp:1699]: BasicBlock::insertInto will make all of these diffs cleaner (supported in 14.X too).

File: src/codegen.cpp [hunk 4]
```diff
@@ -5081,9 +5174,17 @@ static void emit_phinode_assign(jl_codectx_t &ctx, ssize_t idx, jl_value_t *r)
...
             phi->insertAfter(dest);
             PHINode *Tindex_phi = PHINode::Create(getInt8Ty(ctx.builder.getContext()), jl_array_len(edges), "tindex_phi");
+#if JL_LLVM_VERSION >= 160000
+            InsertPt = Tindex_phi->insertInto(BB, InsertPt);
+#else
             BB->getInstList().insert(InsertPt, Tindex_phi);
+#endif
             PHINode *ptr_phi = PHINode::Create(ctx.types().T_prjlvalue, jl_array_len(edges), "ptr_phi");
+#if JL_LLVM_VERSION >= 160000
+            InsertPt = ptr_phi->insertInto(BB, InsertPt);
```
Comment (pchintalapudi) [inline src/codegen.cpp:5184]: This might work unconditionally on 14.X

File: src/llvm-late-gc-lowering.cpp [hunk 5]
```diff
@@ -1600,20 +1600,40 @@ State LateLowerGCFrame::LocalScan(Function &F) {
                         callee->getName() == "memcmp") {
                         continue;
                     }
+#if JL_LLVM_VERSION >= 160000
+                    if (callee->getMemoryEffects().doesNotAccessMemory() ||
+                        callee->getMemoryEffects().onlyReadsMemory() ||
```
Comment (vtjnash) [inline src/llvm-late-gc-lowering.cpp:1605]:
```suggestion
                    if (callee->getMemoryEffects().onlyReadsMemory() ||
```

File: src/llvm-late-gc-lowering.cpp [hunk 6]
```diff
@@ -1600,20 +1600,40 @@ State LateLowerGCFrame::LocalScan(Function &F) {
...
                     if (MemTransferInst *MI = dyn_cast<MemTransferInst>(CI)) {
                         MaybeTrackDst(S, MI);
                     }
                 }
-                if (isa<IntrinsicInst>(CI) || CI->hasFnAttr(Attribute::ArgMemOnly) ||
-                    CI->hasFnAttr(Attribute::ReadNone) || CI->hasFnAttr(Attribute::ReadOnly)) {
+#if JL_LLVM_VERSION >= 160000
+                if (isa<IntrinsicInst>(CI) || 
+                    CI->getMemoryEffects().onlyAccessesArgPointees() ||
+                    CI->getMemoryEffects().doesNotAccessMemory()     ||
```
Comment (vtjnash) [inline src/llvm-late-gc-lowering.cpp:1623]:
```suggestion
```

File: src/pipeline.cpp [hunk 7]
```diff
@@ -351,7 +350,12 @@ static void buildEarlySimplificationPipeline(ModulePassManager &MPM, PassBuilder
         FPM.addPass(DCEPass());
         FPM.addPass(SimplifyCFGPass(basicSimplifyCFGOptions()));
         if (O.getSpeedupLevel() >= 1) {
+#if JL_LLVM_VERSION >= 160000
+            // TODO check the LLVM 15 default.
+            FPM.addPass(SROAPass(SROAOptions::PreserveCFG));
```
Comment (vtjnash) [inline src/pipeline.cpp:355]: The current default is SROAOptions::ModifyCFG, the old default was PreserveCFG

https://github.com/llvm/llvm-project/blob/97b989b27496d5aeadb3b90cbb9305ddcd9e35d2/llvm/include/llvm/Transforms/Scalar.h#L66

https://github.com/llvm/llvm-project/blob/97b989b27496d5aeadb3b90cbb9305ddcd9e35d2/llvm/lib/Passes/PassBuilder.cpp#L1023-L1024

ModifyCFG was added in https://github.com/llvm/llvm-project/commit/4f7e5d22060e8a89237ffb93c3e7be6e92fee8fe#diff-2b09f41dec5c9d1fbe844989c38f6bb7db588799b97b4bed7ad173ce7f348280R1119 for llvm-16

File: src/jitlayers.h [hunk 8]
```diff
@@ -98,7 +102,9 @@ struct OptimizationOptions {
 
 struct NewPM {
     std::unique_ptr<TargetMachine> TM;
+#if JL_LLVM_VERSION < 160000
     StandardInstrumentations SI;
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply src/jitlayers.h:116]: https://reviews.llvm.org/D137149 we can either try to copy the code and maintain our own copies of it, or move this state hidden behind the same ThreadSafeContext as controls the LLVMContext. I think I prefer the latter

[the remark this answers is not in the document]
Comment (vchuravy) [reply src/jitlayers.h:116]: That would mean creating a pass pipeline per module, certainly doable but we should measure the cost that has.

File: src/pipeline.cpp (hunk 7 above)
Comment (vchuravy) [reply src/pipeline.cpp:355]: Thanks for digging that up. So we should use PreserveCFG late in the pipeline and ModifyCFG early on.

File: src/jitlayers.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (vtjnash) [reply src/jitlayers.h:116]: Just per context

File: src/llvm-codegen-shared.h [hunk 9]
```diff
@@ -2,26 +2,21 @@
...
 #include <llvm/Support/ModRef.h>
+#endif
+
 #include "julia.h"
 
 #define STR(csym)           #csym
 #define XSTR(csym)          STR(csym)
 
-#if JL_LLVM_VERSION >= 160000
-namespace llvm {
```
Comment (vchuravy) [inline src/llvm-codegen-shared.h]: @pchintalapudi why did you delete this?

See #51720

Comment (gbaraldi) [reply src/llvm-codegen-shared.h]: I got this working by keeping this define here, making jitlayers.h include codegen-shared and remove the places where we include both (we can also use `#pragma once`)

**Comment by vchuravy** [comment]
This should be NFC for now so let's land this.

**Comment by IanButterworth** [comment]
Build is failing

**Comment by vchuravy** [comment]
> Build is failing

Stupid rebase mistakes.
